torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,203 @@
+            if not isinstance(true_output, torch.Tensor):
+                raise RuntimeError(f"Fallback only works on Tensor output, got output was {type(true_output).__name__}")
+            outputs = tuple(true_output[i] for i in range(true_output.shape[0]))
+        ctx.outputs = outputs
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        outputs = ctx.outputs
+        diff_args = tuple(arg for arg in ctx.args
+                          if isinstance(arg, torch.Tensor) and arg.requires_grad)
+        diff_args = tuple(arg.orig_weight if isinstance(arg, ExpandedWeight) else arg for arg in diff_args)
+        batch_size = grad_output.shape[0]
+        per_sample_grads = tuple(torch.autograd.grad(outputs[i], diff_args, grad_output[i],
+                                                     retain_graph=(i != batch_size - 1))
+                                 for i in range(batch_size))
+        per_sample_grads_iter = zip(*per_sample_grads)
+        result: List[Union[torch.Tensor, None]] = []
+        result.append(None)  # for function input
+        for arg in ctx.args:
+            if isinstance(arg, ExpandedWeight):
+                cur_per_sample_grad = next(per_sample_grads_iter)
+                arg.orig_weight.grad_sample = torch.stack(cur_per_sample_grad)
+                result.append(None)
+            elif isinstance(arg, torch.Tensor) and arg.requires_grad:
+                cur_per_sample_grad = next(per_sample_grads_iter)
+                result.append(torch.stack(cur_per_sample_grad).sum(0))
+            else:
+                result.append(None)
+        result.append(None)
+        return tuple(result)
+
+def forward_helper(func, ctx, expanded_args, num_true_outs):
+    r'''Forward helper computes the forward pass for a function that has expanded weight(s)
+    passed to it. It will run the forward pass where all ExpandedWeights are their original
+    weight. It runs checks on the given arguments and detaches the outputs.
+
+    .. note:: First argument in :attr:`expanded_args` must be the input with the batch
+    dimension as the first element of the shape
+
+    .. note:: :attr:`func` must return a Tensor or tuple of Tensors
+
+    Args:
+        func: The function to be called
+        ctx: The context from the autograd.Function object. Will be used to save
+          computed state from the forward pass
+        expanded_args: Arguments to be passed to :attr:`func`. Will include arguments
+          that need to be unpacked because they are ExpandedWeights
+        num_true_outs: The number of outputs seen by the user since some functions
+          return auxillary data that is only used in the backward pass
+    '''
+    unexpanded_args = _check_and_unexpand_args(func, ctx, expanded_args)
+    output = func(*unexpanded_args)
+    return _check_and_detach_output(ctx, output, num_true_outs)
+
+def _check_and_unexpand_args(func, ctx, expanded_args):
```

Comment:
Does the expanded weights fallback correctly raise an error on operations where we cannot detect if the tensor has a bdim or not? (or is the behavior here that we hope the batch dim is in the correct place?) For example, if the user does something like:

```
torch.add(input, expanded_weight)
```
We probably want to block that operation from happening because we don't know where the batch dim is in the input! It could be in dimension 0, it could be in dimension 1, it might not have one.

Reply:
Currently it doesn't! I think you're right that we should be erroring because of that case but not sure if that means that we should eliminate the fallback entirely then and just error if we can't deal with it. I think we only know the batch dimension location for the `nn.functional` operations, so we could try to look for those operators. Even then, we're assuming that their batch dimension is in dimension 0 which might not be true for all of them

Reply:
I see, maybe we should eliminate the fallback entirely and then just error if we can't deal with it

Reply:
Fallback eliminated and design doc updated